packages/server/src/adapters/aws-lambda.ts

```diff
@@ -0,0 +1,78 @@
+import type { APIGatewayProxyEvent, APIGatewayProxyResult } from 'aws-lambda';
+import { URLSearchParams } from 'url';
+import { resolveHTTPResponse } from '..';
+import { HTTPHeaders, HTTPRequest } from '../http/internals/types';
+import { AnyRouter, inferRouterContext } from '../router';
+import type { AWSLambdaOptions } from './lambda-utils';
+function lambdaEventToHTTPRequest(event: APIGatewayProxyEvent): HTTPRequest {
+  const query = new URLSearchParams();
+  for (const [key, value] of Object.entries(
+    event.multiValueQueryStringParameters ?? {},
+  )) {
+    for (const v of value ?? []) {
+      query.append(key, v);
+    }
+  }
+  return {
+    method: event.httpMethod,
+    query,
+    headers: event.headers,
+    body: event.body,
+  };
+}
+
+export interface CreateLambdaContextOptions {
+  event: APIGatewayProxyEvent;
+}
+
+function transformHeaders(
+  headers: HTTPHeaders,
+): APIGatewayProxyResult['headers'] {
+  const obj: APIGatewayProxyResult['headers'] = {};
+
+  for (const [key, value] of Object.entries(headers)) {
+    if (typeof value === 'undefined') {
+      continue;
+    }
+    obj[key] = Array.isArray(value) ? value.join(', ') : value;
+  }
+  return obj;
+}
+
+export function createApiGatewayHandler<TRouter extends AnyRouter>(
+  opts: AWSLambdaOptions<TRouter, APIGatewayProxyEvent>,
+): (event: APIGatewayProxyEvent) => Promise<APIGatewayProxyResult> {
+  return async (event: APIGatewayProxyEvent) => {
+    const req = lambdaEventToHTTPRequest(event);
+    const path = event.path;
+
+    const createContext = async function _createContext(): Promise<
+      inferRouterContext<TRouter>
+    > {
+      return await opts.createContext?.(event);
+    };
+
+    const response = await resolveHTTPResponse({
+      router: opts.router,
+      batching: opts.batching,
+      responseMeta: opts.responseMeta,
+      createContext,
+      req,
+      path,
+      error: null,
+      onError(o) {
+        opts?.onError?.({
+          ...o,
+          req: event,
```

Comment:
Similar to createContext, I think this should receive `event` rather than `req`

Reply:
Hmm -- this requires some refactoring of internals that'd be hard for @Lilja to do so ignore this and I'll sort it once you're done with the rest